### PR TITLE
Add a verify phase to benchmarks

### PR DESCRIPTION
**Pull request by maysamyabandeh**
Check the result of the benchmark againt a specified truth_db, which is
expected to be produced using the same benchmark but perhaps on a
different commit or with different configs.

The verification is simple and assumes that key/values are generated
deterministically. This assumption would break if db_bench using rand
variable differently from the benchmark that produced truth_db.
Currently it is checked to work on fillrandom and readwhilewriting.

A param finish_after_writes is added to ensure that the background
writing thread will write the same number of entries between two
benchmarks.

Example:
$ TEST_TMPDIR=/dev/shm/truth_db ./db_bench
--benchmarks="fillrandom,readwhilewriting" --num=200000
--finish_after_writes=true
$ TEST_TMPDIR=/dev/shm/tmpdb ./db_bench
--benchmarks="fillrandom,readwhilewriting,verify" --truth_db
/dev/shm/truth_db/dbbench --num=200000 --finish_after_writes=true
Verifying db <= truth_db...
Verifying db >= truth_db...
...Verified

### CODE REVIEW

**Comment by maysamyabandeh** [comment]
Notes: It is not a replacement for stress tests as it does not really stress the db. It also can only catch the bugs only if they persist after the db is reopen. I am thinking that each developer uses that manually for major code changes until it becomes more mature to be used for automated testing.

File: tools/db_bench_tool.cc [hunk 1]
```diff
@@ -2175,6 +2180,37 @@ class Benchmark {
...
+  for (truth_iter->SeekToFirst(); truth_iter->Valid(); truth_iter->Next()) {
+      std::string value;
+      s = db_.db->Get(ro, truth_iter->key(), &value);
+      assert(s.ok());
+      // TODO(myabandeh): provide debugging hints
+      assert(Slice(value) == truth_iter->value());
+  }
+  // Verify that the db iterator does not give any extra key/value
+  fprintf(stderr, "Verifying db >= truth_db...\n");
+  for (db_iter->SeekToFirst(), truth_iter->SeekToFirst(); db_iter->Valid(); db_iter->Next(), truth_iter->Next()) {
```
Comment (sagar0) [inline tools/db_bench_tool.cc]: This loop encompasses the tests in the for-loop between lines 2196-2202, isn't it? So, curious why the previous loop is needed to test using point lookups (unless you specifically want to test point lookups).

File: tools/db_bench_tool.cc [hunk 2]
```diff
@@ -2175,6 +2180,37 @@ class Benchmark {
...
+  if (!s.ok()) {
+    fprintf(stderr, "open error: %s\n", s.ToString().c_str());
+    exit(1);
+  }
+  ReadOptions ro;
+  ro.total_order_seek = true;
+  std::unique_ptr<Iterator> truth_iter(truth_db.db->NewIterator(ro));
+  std::unique_ptr<Iterator> db_iter(db_.db->NewIterator(ro));
+  // Verify that all the key/values in truth_db are retrivable in db with ::Get
+  fprintf(stderr, "Verifying db <= truth_db...\n");
```
Comment (sagar0) [inline tools/db_bench_tool.cc]: What do you mean by "db <= truth_db" here.
May be I don't fully understand, but I believe it should be db >= truth_db here, and db == truth_db in line 2204.

Comment (maysamyabandeh) [inline tools/db_bench_tool.cc]: Oh, yes you are correct.

File: tools/db_bench_tool.cc (hunk 1 above)
Comment (maysamyabandeh) [inline tools/db_bench_tool.cc]: Yeah the previous loop was to test ::Get

**Pull request closed without merging**